### PR TITLE
JCR-4608 Upgrade to commons-collections4

### PR DESCRIPTION
**Pull request by rkovarik**
https://issues.apache.org/jira/browse/JCR-4608
commons-collections 3.2.2 suffers from https://advisory.checkmarx.net/advisory/vulnerability/Cx78f40514-81ff/
Although Jackrabbit itself does not seem to use SetUniqueList, it still exists on the classpath.

### CODE REVIEW

**Comment by reschke** [comment]
can you try to rebase this based on what's currently in trunk?

File: jackrabbit-jcr2spi/src/main/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImpl.java [hunk 1]
```diff
@@ -704,15 +701,10 @@ public LinkedEntries.LinkNode next() {
                 return n;
             }
 
+            @Override
             public void remove() {
                 throw new UnsupportedOperationException("remove");
             }
-
```
Comment (reschke) [inline jackrabbit-jcr2spi/src/main/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImpl.java]: I stumbled across this as well; the original code apparently tried to protect itself against modifcations through other methods. Can we really eliminate this?

Comment (rkovarik) [reply jackrabbit-jcr2spi/src/main/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImpl.java]: I've added a test for concurrent modification. Was passing when I accessed the original modCount field via reflection but doesn't work when I overlay the field. Must be a stupid mistake but I can't see it 🤔

File: jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.Test;
+
+public class ChildNodeEntriesImplTest {
+
+    private final ExecutorService executorService = Executors.newFixedThreadPool(10);
+
+    @Test(expected = ConcurrentModificationException.class)
+    public void concurrentModification() throws Throwable {
+        ChildNodeEntriesImpl entries = new ChildNodeEntriesImpl(null, null, null);
+        Collection<Future<?>> futures = new ArrayList<>();
```
Comment (mduerig) [inline jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: @rkovarik the test failing looks like a timing issue. If you pre-populate the child nodes entries it has a greater chance to actually run into a ConcurrentModificationException. 
I.e. adding 100 child nodes did the trick for me. 

        for (int k = 0; k < 100; k++) {
            entries.add(mock(NodeEntry.class));
        }

Comment (rkovarik) [reply jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: Right, thx @mduerig. I get only up to 1000 successful runs so I will remove the test, the code looks fine.

Comment (reschke) [reply jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: I'd still like to see a test (that we can make fail by removing the modified checks in the old code) before proceeding.

Comment (rkovarik) [reply jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: OK, pushed it back, one timing failure in ~10 000 runs. https://github.com/apache/jackrabbit/pull/128/commits/fab272397237d4e71bd1e3fc0a8df87206c61d7b
Screen Shot 2022-10-05 at 9 52 49

Comment (mduerig) [reply jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: The test is non deterministic so I'd expect it to fail from time to time. I don't see a way to implement a deterministic test here unless we make `LinkedEntries` package private and unit test that class directly.

Comment (reschke) [reply jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java]: I already moved it into a separate class (see https://issues.apache.org/jira/browse/JCR-4854)

**Comment by reschke** [comment]
@rkovarik , @mduerig  - please see https://issues.apache.org/jira/browse/JCR-4862 (which is the last non-trivial step in removing the old dependency)

**Comment by reschke** [comment]
(OBE)

**Pull request closed without merging**